core/modules/css/KvInterfaceImplMySql.cc

```diff
@@ -0,0 +1,503 @@
+        LOGF(_logger, LOG_LVL_ERROR, "SQL DELETE FROM failed (too many (%1%) rows deleted): %2%" % affectedRows % query);
+        throw CssError("deleteKey - unexpectedly deleted more than 1 row.");
+    }
+
+}
+
+
+std::string
+KvInterfaceImplMySql::_get(std::string const& key, std::string const& defaultValue, bool throwIfKeyNotFound) {
+    // Note regarding starting a transaction here:
+    // Most private methods (in fact all of them) require that a transaction already be in progress. On the other hand:
+    // this method is protected, and is called by KvInterface with no transaction in progress so here one must be started.
+    KvTransaction transaction(_conn);
+
+    std::string val;
+    sql::SqlErrorObject errObj;
+
+    boost::format fmQuery("SELECT kvVal FROM kvData WHERE kvKey='%1%'");
+    fmQuery % key;
+    std::string query = fmQuery.str();
+
+    sql::SqlResults results;
+    if (not _conn.runQuery(query, results, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "_get - query failed: %1%" % query);
+        if (throwIfKeyNotFound) {
+            throw CssError(errObj);
+        } else {
+            val = defaultValue;
+        }
+    } else {
+        errObj.reset();
+        if (not results.extractFirstValue(val, errObj)) {
+            if (throwIfKeyNotFound) {
+                throw NoSuchKey(errObj);
+            } else {
+                val = defaultValue;
+            }
+        }
+    }
+
+    transaction.commit();
+    return val;
+}
+
+
+bool KvInterfaceImplMySql::_getIdFromServer(std::string const& key, unsigned int* id) {
+    // a transaction should be in process here.
+
+    boost::format fmQuery("SELECT kvId FROM kvData WHERE kvKey='%1%'");
+    fmQuery % key;
+    std::string query = fmQuery.str();
+
+    sql::SqlResults results;
+    sql::SqlErrorObject errObj;
+    if (not _conn.runQuery(query, results, errObj)) {
+        LOGF(_logger, LOG_LVL_ERROR, "_getIdFromServer - query failed: %1%" % query);
+        return false;
+    }
+    errObj.reset();
+
+    if (not extractIntValueFromSqlResults(&results, id)) {
+        return false;
+    }
+
+    return true;
+}
+
+
+std::string KvInterfaceImplMySql::_validateKey(std::string const& key) {
+    // There is no need for a transaction here.
+
+    sql::SqlErrorObject errObj;
+    std::string retStr;
+    if (not _conn.escapeString(key, retStr, errObj)) {
+        throw CssError(errObj);
+    }
+    // verify that:
+    // - key is less than max length
+    // - first character should be a delimiter
+    // - last character should not be a delimiter
+    if (key.size() > KvInterface::MAX_KEY_LENGTH ||
+            key.find_first_of(KEY_PATH_DELIMITER) != 0 ||
+            key.find_last_of(KEY_PATH_DELIMITER) == key.size()-1) {
+        LOGF(_logger, LOG_LVL_DEBUG, "create - rejecting key: %1%" % retStr);
+        throw CssError("invalid key");
+    }
+    return retStr;
+}
+
+
+std::string KvInterfaceImplMySql::_validateValue(std::string const& value) {
```

Comment:
OTOH we might want to add validation to the value, which is why I put it in and named it thus. Otherwise it'd be as easy to just do the escaping inline I think.